include/pybind11/pybind11.h

```diff
@@ -493,7 +493,11 @@ class cpp_function : public function {
                 } else {
                     std::string tname(t->name());
                     detail::clean_type_id(tname);
-                    signature += tname;
+                    if (detail::cpp_name_needs_typing_annotated(tname.c_str())) {
+                        signature += "Annotated[Any, CppTypePybind11(\"" + tname + "\")]";
+                    } else {
+                        signature += tname;
+                    }
```

Comment:
Strictly speaking, we should always wrap C++ types with `Annotated` at this point. 
Being a top-level non-template type is not enough to assume 1-to-1 C++-Python type name correspondence, although it is the case for (some) built-in types.

Reply:
Adding the following to `cast.h` allows us to remove the above condition without turning all tests red.

```C++
template <>
struct handle_type_name<str> {
    static constexpr auto name = const_name("str");
};
template <>
struct handle_type_name<set> {
    static constexpr auto name = const_name("set");
};
template <>
struct handle_type_name<anyset> {
    static constexpr auto name = const_name("set");
};
template <>
struct handle_type_name<tuple> {
    static constexpr auto name = const_name("tuple");
};
template <>
struct handle_type_name<frozenset> {
    static constexpr auto name = const_name("frozenset");
};
template <>
struct handle_type_name<list> {
    static constexpr auto name = const_name("list");
};
template <>
struct handle_type_name<dict> {
    static constexpr auto name = const_name("dict");
};
template <>
struct handle_type_name<object> {
    static constexpr auto name = const_name("object");
};
```


The only two failures probably indicate the error in tests themselves:

<details><summary>Pytest errors</summary>
<p>


```
====================================================================================================================== FAILURES ======================================================================================================================
___________________________________________________________________________________________________________________ test_signature ___________________________________________________________________________________________________________________

doc = <conftest.SanitizedString object at 0x7fafa7b3f8e0>

    def test_signature(doc):
>       assert (
            doc(m.create_rec_nested)
            == "create_rec_nested(arg0: int) -> numpy.ndarray[NestedStruct]"
        )
E       assert --- actual / +++ expected
E         - create_rec_nested(arg0: int) -> numpy.ndarray[Annotated[Any, CppTypePybind11("NestedStruct")]]
E         ?                                               --------------------------------            -- -
E         + create_rec_nested(arg0: int) -> numpy.ndarray[NestedStruct]

doc        = <conftest.SanitizedString object at 0x7fafa7b3f8e0>

../../tests/test_numpy_dtypes.py:344: AssertionError
______________________________________________________________________________________________________________________ test_set ______________________________________________________________________________________________________________________

capture = <conftest.Capture object at 0x7fafa82995a0>, doc = <conftest.SanitizedString object at 0x7fafa82991b0>

    def test_set(capture, doc):
        s = m.get_set()
        assert isinstance(s, set)
        assert s == {"key1", "key2", "key3"}
    
        s.add("key4")
        with capture:
            m.print_anyset(s)
        assert (
            capture.unordered
            == """
            key: key1
            key: key2
            key: key3
            key: key4
        """
        )
    
        m.set_add(s, "key5")
        assert m.anyset_size(s) == 5
    
        m.set_clear(s)
        assert m.anyset_empty(s)
    
        assert not m.anyset_contains(set(), 42)
        assert m.anyset_contains({42}, 42)
        assert m.anyset_contains({"foo"}, "foo")
    
        assert doc(m.get_set) == "get_set() -> set"
>       assert doc(m.print_anyset) == "print_anyset(arg0: anyset) -> None"
E       assert --- actual / +++ expected
E         - print_anyset(arg0: set) -> None
E         + print_anyset(arg0: anyset) -> None
E         ?                    +++

capture    = <conftest.Capture object at 0x7fafa82995a0>
doc        = <conftest.SanitizedString object at 0x7fafa82991b0>
s          = set()
```


</p>
</details>

Reply:
Sorry I didn't point out before, I got the same result already, it's in a commit here: 76b4a34aff6e7fd1ae2728588c5fd5914f7320d0

The only remaining errors are related to numpy `NestedStruct`. I tried to pin-point where the `%` originates, but I gave up after looking for ~5 minutes, reverted the commit, and then tried the `cpp_name_needs_typing_annotated()` heuristic instead to see how far we get with that.

I agree what you suggest here is best, but someone needs to figure out how to deal with the failing test, and then we have to decide if we need to roll out the behavior change in stages, maybe opt-in first in one release, then opt-out in the next.

Reply:
Oh, right. You already tried that. I should have noticed.

I think we should apply the following diff to tests. 
 - `NestedStruct` doesn't have a descriptive python counterpart, the `CppTypePybind11("NestedType")` is the best we can do.
 - There is no such thing as  `anyset` in python.
```c++
template <>
struct handle_type_name<anyset> {
    static constexpr auto name = const_name("set");
};
```

This should resolve the troubles with broken tests.

<details><summary>diff</summary>
<p>

```diff
diff --git a/tests/test_numpy_dtypes.py b/tests/test_numpy_dtypes.py
index d10457ee..2fc3771c 100644
--- a/tests/test_numpy_dtypes.py
+++ b/tests/test_numpy_dtypes.py
@@ -343,7 +343,7 @@ def test_complex_array():
 def test_signature(doc):
     assert (
         doc(m.create_rec_nested)
-        == "create_rec_nested(arg0: int) -> numpy.ndarray[NestedStruct]"
+        == "create_rec_nested(arg0: int) -> numpy.ndarray[Annotated[Any, CppTypePybind11(\"NestedStruct\")]]"
     )
 
 
diff --git a/tests/test_pytypes.py b/tests/test_pytypes.py
index 2b202731..2824d29f 100644
--- a/tests/test_pytypes.py
+++ b/tests/test_pytypes.py
@@ -121,7 +121,7 @@ def test_set(capture, doc):
     assert m.anyset_contains({"foo"}, "foo")
 
     assert doc(m.get_set) == "get_set() -> set"
-    assert doc(m.print_anyset) == "print_anyset(arg0: anyset) -> None"
+    assert doc(m.print_anyset) == "print_anyset(arg0: set) -> None"
```

</p>
</details>

Reply:
> `NestedStruct` doesn't have a descriptive python counterpart

Nice, thanks! I didn't know / wasn't sure before. I'm glad I gave up before, finding out the hard way. :-)

> There is no such thing as `anyset` in python.

Of course! :-) Thanks for catching that as well, I was just going through with rushed lawnmower mentality.

Please take another look, the `cpp_name_needs_typing_annotated()` heuristic is gone, all tests pass (locally at least).